pkg/prove/proof.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/tendermint/tendermint/types"
+)
+
+const (
+	usableShareSize = consts.ShareSize - consts.NamespaceSize - consts.ShareReservedBytes
+)
+
+// TxInclusion uses the provided block data to progressively generate rows
+// of a data square, and then using those shares to creates nmt inclusion proofs
+// It is possible that a transaction spans more than one row. In that case, we
+// have to return two proofs.
+func TxInclusion(codec rsmt2d.Codec, data types.Data, origSquareSize, txIndex int) (types.TxProof, error) {
+	startPos, endPos := txSharePosition(data.Txs, txIndex)
+	startRow := startPos / origSquareSize
+	endRow := endPos / origSquareSize
+
+	if (endPos - startPos) > 1 {
+		return types.TxProof{}, errors.New("transaction spanned more than two shares, this is not yet supported")
+	}
+
+	rowShares := genRowShares(consts.DefaultCodec(), data, origSquareSize, startRow, endRow)
+
+	var proofs []*tmproto.NMTProof  //nolint:prealloc // rarely will this contain more than a single proof
+	var shares [][]byte             //nolint:prealloc // rarely will this contain more than a single share
+	var rowRoots []tmbytes.HexBytes //nolint:prealloc // rarely will this contain more than a single root
+	for i, row := range rowShares {
+		// create an nmt to use to generate a proof
+		tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(origSquareSize))
+		for j, share := range row {
+			tree.Push(
+				share,
+				rsmt2d.SquareIndex{
+					Axis: uint(i),
+					Cell: uint(j),
+				},
+			)
+		}
+
+		var pos int
+		if i == 0 {
+			pos = startPos - (startRow * origSquareSize)
+		} else {
+			pos = endPos - (endRow * origSquareSize)
+		}
+
+		shares = append(shares, row[pos])
+
+		proof, err := tree.Prove(pos)
+		if err != nil {
+			return types.TxProof{}, err
+		}
+
+		proofs = append(proofs, &tmproto.NMTProof{
+			Start:    int32(proof.Start()),
+			End:      int32(proof.End()),
+			Nodes:    proof.Nodes(),
+			LeafHash: proof.LeafHash(),
+		})
+
+		// we don't store the data availability header anywhere, so we
+		// regenerate the roots to each row
+		rowRoots = append(rowRoots, tree.Root())
+	}
+
+	return types.TxProof{
+		RowRoots: rowRoots,
+		Data:     shares,
+		Proofs:   proofs,
+	}, nil
+}
+
+// txSharePosition returns the share that a given transaction is included in.
+// returns -1 if index is greater than that of the provided txs.
+func txSharePosition(txs types.Txs, txIndex int) (startSharePos, endSharePos int) {
+	if txIndex >= len(txs) {
+		return -1, -1
+	}
+
+	totalLen := 0
+	for i := 0; i < txIndex; i++ {
+		txLen := len(txs[i])
+		totalLen += (delimLen(txLen) + txLen)
+	}
+
+	txLen := len(txs[txIndex])
+
+	startSharePos = (totalLen) / usableShareSize
+	endSharePos = (totalLen + txLen + delimLen(txLen)) / usableShareSize
+
+	return startSharePos, endSharePos
+}
+
+func delimLen(txLen int) int {
+	lenBuf := make([]byte, binary.MaxVarintLen64)
+	return binary.PutUvarint(lenBuf, uint64(txLen))
+}
```

Comment:
This looks like a re-implementation of length-delimited protobuf marshaling. Can't you just use that instead, then take the number of bytes of the marshaled data?

Reply:
it is, and we could! However, in order not to marshal the txs more than once, we would have to refactor to 
- split txs into shares
- find our tx
- possibly generate the rest of the row
- generate proof

as opposed to how we do this now
- calculate tx position
- generate necessary shares
- generate proof

happy to refactor if we think it's worth getting rid of this function